### PR TITLE
Give up the use of fixed gpio numbers

### PR DESCRIPTION
**Pull request by AsuraZeng**
insead of using dynamic gpio numbers according to the probe sequence

Signed-off-by: Chao Zeng <chao.zeng@siemens.com>

### CODE REVIEW

**Comment by AsuraZeng** [comment]
these changes are for upstream kernel.
should be used when migrating to the upstream kernel.

**Comment by jan-kiszka** [comment]
> these changes are for upstream kernel.
> should be used when migrating to the upstream kernel.

But they do not break the 4.19-ti kernel, do they?

**Comment by AsuraZeng** [comment]
This means that it cannot be used with the current ti kernel. To be used with the upstream kernel

**Comment by jan-kiszka** [comment]
Because wkup_gpio0_base does not match the DT used there? Maybe you can add the corresponding patch for the TI kernel as well and probe for what is available? Then we can easily allow switching kernels without patching userland.

**Comment by AsuraZeng** [comment]
I think just like MRAA, there should be some modifications for the switchserial on the userland instead of patching for the ti kernel.

**Comment by jan-kiszka** [comment]
I'm not suggesting to patch the kernel. I'm suggesting to account for any potential differences between upstream DT and TI DT in the switch-serial tool to make it work on both.

**Comment by BaochengSu** [comment]
Should we use the "gpiolib" interface introduced since v4.8 kernel?

**Comment by jan-kiszka** [comment]
If it simplifies the code, let's do that. But speed is surely not a driver here.

**Comment by gramss** [comment]
> 
> 
> Should we use the "gpiolib" interface introduced since v4.8 kernel?

As suggested here: #40 , I can only say +1 on that.. 👍

File: recipes-app/switchserialmode/files/src/switchserialmode.c [hunk 1]
```diff
@@ -237,20 +297,56 @@ void gpio_set(const char *gpiodev, const char *offset, const char *value)
...
+    int offset = 4;//4 5 6 7
+    char uart_path[MAX_SIZE] = {0};
+    char uart_base[MAX_SIZE] = {0};
+    int num;
+    int wkup_gpio0_base = get_gpiochip_base("/sys/devices/platform/interconnect@100000/interconnect@100000:interconnect@28380000/interconnect@100000:interconnect@28380000:interconnect@42040000/42110000.wkup_gpio0");
+
+    if(wkup_gpio0_base < 0)
+    {
+        return -1;
+    }
```
Comment (jan-kiszka) [inline recipes-app/switchserialmode/files/src/switchserialmode.c:312]: Inconsistent coding style.

File: recipes-app/switchserialmode/files/src/switchserialmode.c [hunk 2]
```diff
@@ -237,20 +297,56 @@ void gpio_set(const char *gpiodev, const char *offset, const char *value)
...
+    gpio_set(ctr_info.uart_ctrl_path[UART_EN_CTRL],ctr_info.uart_ctrl_base[UART_EN_CTRL] , "1");            /* uart_en gpio */
+    gpio_set(ctr_info.uart_ctrl_path[UART_MODE0_CTRL], ctr_info.uart_ctrl_base[UART_MODE0_CTRL], va ? "1" : "0"); /* uart_mode0 gpio */
+    gpio_set(ctr_info.uart_ctrl_path[UART_MODE1_CTRL], ctr_info.uart_ctrl_base[UART_MODE1_CTRL], vb ? "1" : "0"); /* uart_mode1 gpio */
 }
 
 static void gpio_set_terminate(int onoff)
 {
-    gpio_set("/sys/class/gpio/gpio193", "193", onoff ? "1" : "0"); /* uart_terminate gpio */
+    gpio_set(ctr_info.uart_ctrl_path[UART_TERMINATE_CTRL], ctr_info.uart_ctrl_base[UART_TERMINATE_CTRL], onoff ? "1" : "0"); /* uart_terminate gpio */
+
```
Comment (jan-kiszka) [inline recipes-app/switchserialmode/files/src/switchserialmode.c:339]: stray newline

File: recipes-app/switchserialmode/files/src/switchserialmode.c [hunk 3]
```diff
@@ -237,20 +297,56 @@ void gpio_set(const char *gpiodev, const char *offset, const char *value)
     exe_shell(cmd_set_gpio_value, LEN(cmd_set_gpio_value), NULL, 0);
 }
 
+static int uart_ctrl_info_init(void)
+{
+    int i;
+    int offset = 4;//4 5 6 7
```
Comment (jan-kiszka) [inline recipes-app/switchserialmode/files/src/switchserialmode.c:303]: Forgotten debug values?

File: recipes-app/switchserialmode/files/src/switchserialmode.c [hunk 4]
```diff
@@ -221,6 +244,43 @@ int exe_shell(const char **lines, int lines_num, char *res, int res_size)
...
+    snprintf(buf, sizeof(buf) - 1, "%s/gpio/gpiochip*/base", device_path);
+    base_file = file_unglob(buf);
+    if (!base_file)
+        return -1;
+
+    fh = fopen(base_file, "r");
+    if (!fh)
+        return -1;
+
+    ret = fread(buf, 1, sizeof(buf) - 1, fh);
```
Comment (jan-kiszka) [inline recipes-app/switchserialmode/files/src/switchserialmode.c:274]: you could also use fscanf.

**Comment by jan-kiszka** [comment]
This became obsolete via #95.

**Pull request closed without merging**